File: sdks/java/aether-client/src/test/java/io/aether/sdk/client/IntegrationTest.java
package io.aether.sdk.client;

import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.condition.DisabledIfEnvironmentVariable;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Integration tests for the Java SDK against a running Aether server.
 *
 * <p>These tests require a running Aether server at the AETHER_BASE_URL
 * (default: http://localhost:8080). They are skipped automatically if
 * the server is not reachable.</p>
 *
 * <p>Run with:
 * <pre>{@code
 * mvn verify -pl aether-client
 * AETHER_BASE_URL=http://localhost:9090 mvn verify -pl aether-client
 * }</pre></p>
 */
@DisabledIfEnvironmentVariable(named = "AETHER_SKIP_INTEGRATION", matches = "true")
class IntegrationTest {

    private static final String BASE_URL = System.getenv().getOrDefault("AETHER_BASE_URL", "http://localhost:8080");
    private static boolean serverReachable;

    @BeforeAll
    static void checkServer() {
        try {
            HttpClient client = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(3)).build();
            HttpResponse<String> resp = client.send(
                    HttpRequest.newBuilder().uri(URI.create(BASE_URL + "/health")).GET().build(),
                    HttpResponse.BodyHandlers.ofString()
            );
            serverReachable = resp.statusCode() == 200;
        } catch (Exception e) {
            serverReachable = false;
        }
        org.junit.jupiter.api.Assumptions.assumeTrue(serverReachable,
                "Aether server not reachable at " + BASE_URL);
    }

    private static String uniqueId() {
        return "inttest-" + UUID.randomUUID().toString().replace("-", "").substring(0, 12);
    }

    @Nested
    @DisplayName("Health Check")
    class HealthCheck {

        @Test
        @DisplayName("health returns ok status")
        void healthReturnsOk() {
            try (AetherClient client = AetherClient.builder(BASE_URL).build()) {
                AetherClient.ServerInfo info = client.health();
                assertEquals("ok", info.getStatus());
                assertTrue(info.getUptime() >= 0);
            }
        }

        @Test
        @DisplayName("info returns version or counts")
        void infoReturnsData() {
            try (AetherClient client = AetherClient.builder(BASE_URL).build()) {
                Map<String, Object> info = client.info();
                assertTrue(info.containsKey("version") || info.containsKey("actor_count"),
                        "expected 'version' or 'actor_count' in info response");
            }
        }
    }

    @Nested
    @DisplayName("Actor Spawn and Message Send")
    class ActorSpawnAndMessage {

        @Test
        @DisplayName("register actor returns correct info")
        void registerActor() {
            try (AetherClient client = AetherClient.builder(BASE_URL).build()) {
                String actorId = uniqueId();
                AetherClient.ActorInfo info = client.registerActor(actorId, "worker", List.of(), Map.of());
                assertEquals(actorId, info.getActorId());
                assertEquals("worker", info.getActorType());
                assertEquals("active", info.getStatus());
            }
        }

        @Test
        @DisplayName("get actor returns correct info")
        void getActor() {
            try (AetherClient client = AetherClient.builder(BASE_URL).build()) {
                String actorId = uniqueId();
                client.registerActor(actorId, "worker", List.of(), Map.of());
                AetherClient.ActorInfo info = client.getActor(actorId);
                assertEquals(actorId, info.getActorId());
            }
        }

        @Test
        @DisplayName("list actors includes registered actor")
        void listActors() {
            try (AetherClient client = AetherClient.builder(BASE_URL).build()) {
                String actorId = uniqueId();
                client.registerActor(actorId, "integration-test", List.of(), Map.of());
                List<AetherClient.ActorInfo> actors = client.listActors();
                boolean found = actors.stream().anyMatch(a -> a.getActorId().equals(actorId));
                assertTrue(found, "registered actor not found in list");
            }
        }

        @Test
        @DisplayName("unregister actor then get returns error")
        void unregisterActor() {
            try (AetherClient client = AetherClient.builder(BASE_URL).build()) {
                String actorId = uniqueId();
                client.registerActor(actorId, "worker", List.of(), Map.of());
                client.unregisterActor(actorId);
                assertThrows(AetherServerError.class, () -> client.getActor(actorId));
            }
        }

        @Test
        @DisplayName("send message returns delivered receipt")
        void sendMessage() {
            try (AetherClient client = AetherClient.builder(BASE_URL)
                    .actorId("integration-test-sender").build()) {
                String actorId = uniqueId();
                client.registerActor(actorId, "worker", List.of(), Map.of());
                AetherClient.DeliveryReceipt receipt = client.sendMessage(
                        actorId, Map.of("hello", "world"));
                assertEquals("delivered", receipt.getStatus());
                assertNotNull(receipt.getMessageId());
            }
        }

        @Test
        @DisplayName("heartbeat succeeds for registered actor")
        void heartbeat() {
            try (AetherClient client = AetherClient.builder(BASE_URL).build()) {
                String actorId = uniqueId();
                client.registerActor(actorId, "worker", List.of(), Map.of());
                assertDoesNotThrow(() -> client.heartbeat(actorId));
            }
        }
    }

    @Nested
    @DisplayName("State Read/Write")
    class StateReadWrite {

        @Test
        @DisplayName("set and get state round-trips correctly")
        void setAndGetState() {
            try (AetherClient client = AetherClient.builder(BASE_URL).build()) {
                String actorId = uniqueId();
                client.registerActor(actorId, "worker", List.of(), Map.of());
                client.setState(actorId, "counter", 42);
                Optional<AetherClient.StateEntry> entry = client.getState(actorId, "counter");
                assertTrue(entry.isPresent());
                assertEquals(42, entry.get().getValue());
            }
        }

        @Test
        @DisplayName("get missing state returns empty")
        void getMissingState() {
            try (AetherClient client = AetherClient.builder(BASE_URL).build()) {
                String actorId = uniqueId();
                client.registerActor(actorId, "worker", List.of(), Map.of());
                Optional<AetherClient.StateEntry> entry = client.getState(actorId, "nonexistent");
                assertTrue(entry.isEmpty());
            }
        }

        @Test
        @DisplayName("delete state returns true for existing key")
        void deleteState() {
            try (AetherClient client = AetherClient.builder(BASE_URL).build()) {
                String actorId = uniqueId();
                client.registerActor(actorId, "worker", List.of(), Map.of());
                client.setState(actorId, "temp", "data");
                assertTrue(client.deleteState(actorId, "temp"));
            }
        }

        @Test
        @DisplayName("get all state returns written keys")
        void getAllState() {
            try (AetherClient client = AetherClient.builder(BASE_URL).build()) {
                String actorId = uniqueId();
                client.registerActor(actorId, "worker", List.of(), Map.of());
                client.setState(actorId, "a", 1);
                client.setState(actorId, "b", 2);
                Map<String, Object> state = client.getAllState(actorId);
                assertTrue(state.containsKey("a") || state.containsKey("b"));
            }
        }

        @Test
        @DisplayName("state version increments on update")
        void stateVersionIncrements() {
            try (AetherClient client = AetherClient.builder(BASE_URL).build()) {
                String actorId = uniqueId();
                client.registerActor(actorId, "worker", List.of(), Map.of());
                AetherClient.StateEntry e1 = client.setState(actorId, "counter", 1);
                AetherClient.StateEntry e2 = client.setState(actorId, "counter", 2);
                assertTrue(e2.getVersion() > e1.getVersion(),
                        "expected version to increment: " + e1.getVersion() + " -> " + e2.getVersion());
            }
        }
    }

    @Nested
    @DisplayName("Mesh Peer Discovery")
    class MeshPeerDiscovery {

        @Test
        @DisplayName("cluster info endpoint returns data")
        void clusterInfo() {
            try {
                HttpClient http = HttpClient.newBuilder()
                        .connectTimeout(Duration.ofSeconds(5)).build();
                HttpResponse<String> resp = http.send(
                        HttpRequest.newBuilder().uri(URI.create(BASE_URL + "/cluster/info")).GET().build(),
                        HttpResponse.BodyHandlers.ofString());
                if (resp.statusCode() == 404) {
                    org.junit.jupiter.api.Assumptions.assumeTrue(false,
                            "Cluster endpoints not available on this server");
                }
                assertEquals(200, resp.statusCode());
                assertTrue(resp.body().contains("node_id")
                                || resp.body().contains("cluster_enabled")
                                || resp.body().contains("status"),
                        "expected cluster info fields in response");
            } catch (IOException | InterruptedException e) {
                org.junit.jupiter.api.Assumptions.assumeTrue(false,
                        "Cluster endpoint not available: " + e.getMessage());
            }
        }

        @Test
        @DisplayName("cluster nodes endpoint returns data")
        void clusterNodes() {
            try {
                HttpClient http = HttpClient.newBuilder()
                        .connectTimeout(Duration.ofSeconds(5)).build();
                HttpResponse<String> resp = http.send(
                        HttpRequest.newBuilder().uri(URI.create(BASE_URL + "/cluster/nodes")).GET().build(),
                        HttpResponse.BodyHandlers.ofString());
                if (resp.statusCode() == 404) {
                    org.junit.jupiter.api.Assumptions.assumeTrue(false,
                            "Cluster endpoints not available on this server");
                }
                assertEquals(200, resp.statusCode());
            } catch (IOException | InterruptedException e) {
                org.junit.jupiter.api.Assumptions.assumeTrue(false,
                        "Cluster endpoint not available: " + e.getMessage());
            }
        }
    }
}

File: sdks/go/tests/integration_test.go
package aether_test

import (
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"strings"
	"testing"
	"time"

	"github.com/WyattAu/aether-core/sdks/go/aether"
)

var baseURL = "http://localhost:8080"

func init() {
	if u := os.Getenv("AETHER_BASE_URL"); u != "" {
		baseURL = strings.TrimRight(u, "/")
	}
}

func serverReachable() bool {
	client := &http.Client{Timeout: 3 * time.Second}
	resp, err := client.Get(baseURL + "/health")
	if err != nil {
		return false
	}
	defer resp.Body.Close()
	return resp.StatusCode == 200
}

func needsServer(t *testing.T) {
	t.Helper()
	if !serverReachable() {
		t.Skipf("Aether server not reachable at %s", baseURL)
	}
}

func uniqueID(prefix string) string {
	return fmt.Sprintf("%s-%d", prefix, time.Now().UnixNano())
}

func TestIntegrationHealthCheck(t *testing.T) {
	needsServer(t)

	client := aether.NewClient(baseURL)
	info, err := client.Health()
	if err != nil {
		t.Fatalf("health check failed: %v", err)
	}

	if info.Uptime < 0 {
		t.Errorf("expected uptime >= 0, got %f", info.Uptime)
	}
}

func TestIntegrationInfoEndpoint(t *testing.T) {
	needsServer(t)

	httpClient := &http.Client{Timeout: 5 * time.Second}
	resp, err := httpClient.Get(baseURL + "/api/v1/info")
	if err != nil {
		t.Fatalf("info request failed: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		t.Fatalf("expected 200, got %d", resp.StatusCode)
	}

	var data map[string]any
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		t.Fatalf("failed to decode info: %v", err)
	}

	if _, ok := data["version"]; !ok {
		if _, ok := data["actor_count"]; !ok {
			t.Error("expected 'version' or 'actor_count' in info response")
		}
	}
}

func TestIntegrationRegisterActor(t *testing.T) {
	needsServer(t)

	client := aether.NewClient(baseURL)
	actorID := uniqueID("inttest")

	info, err := client.RegisterActor(actorID, "worker", []string{"compute"}, map[string]any{"region": "us-east"})
	if err != nil {
		t.Fatalf("register actor failed: %v", err)
	}

	if info.ActorID != actorID {
		t.Errorf("expected actor_id %s, got %s", actorID, info.ActorID)
	}
	if info.ActorType != "worker" {
		t.Errorf("expected actor_type worker, got %s", info.ActorType)
	}
}

func TestIntegrationGetActor(t *testing.T) {
	needsServer(t)

	client := aether.NewClient(baseURL)
	actorID := uniqueID("inttest")

	_, err := client.RegisterActor(actorID, "worker", nil, nil)
	if err != nil {
		t.Fatalf("register actor failed: %v", err)
	}

	got, err := client.GetActor(actorID)
	if err != nil {
		t.Fatalf("get actor failed: %v", err)
	}

	if got.ActorID != actorID {
		t.Errorf("expected %s, got %s", actorID, got.ActorID)
	}
}

func TestIntegrationListActors(t *testing.T) {
	needsServer(t)

	client := aether.NewClient(baseURL)
	actorID := uniqueID("inttest")

	_, err := client.RegisterActor(actorID, "integration-test", nil, nil)
	if err != nil {
		t.Fatalf("register actor failed: %v", err)
	}

	actors, err := client.ListActors("", "")
	if err != nil {
		t.Fatalf("list actors failed: %v", err)
	}

	found := false
	for _, a := range actors {
		if a.ActorID == actorID {
			found = true
			break
		}
	}
	if !found {
		t.Errorf("actor %s not found in list", actorID)
	}
}

func TestIntegrationUnregisterActor(t *testing.T) {
	needsServer(t)

	client := aether.NewClient(baseURL)
	actorID := uniqueID("inttest")

	client.RegisterActor(actorID, "worker", nil, nil)

	err := client.UnregisterActor(actorID)
	if err != nil {
		t.Fatalf("unregister actor failed: %v", err)
	}

	_, err = client.GetActor(actorID)
	if err == nil {
		t.Error("expected error when getting unregistered actor")
	}
}

func TestIntegrationSendMessage(t *testing.T) {
	needsServer(t)

	client := aether.NewClient(baseURL, aether.WithActorID("integration-test-sender"))
	actorID := uniqueID("inttest")

	_, err := client.RegisterActor(actorID, "worker", nil, nil)
	if err != nil {
		t.Fatalf("register actor failed: %v", err)
	}

	envelope := &aether.MessageEnvelope{
		SourceActor: "integration-test-sender",
		TargetActor: actorID,
		MessageType: "greeting",
		Payload:     map[string]string{"hello": "world"},
	}

	receipt, err := client.SendMessage(actorID, envelope)
	if err != nil {
		t.Fatalf("send message failed: %v", err)
	}

	if receipt.Status != "delivered" {
		t.Errorf("expected status 'delivered', got '%s'", receipt.Status)
	}
	if receipt.MessageID == "" {
		t.Error("expected non-empty message_id")
	}
}

func TestIntegrationHeartbeat(t *testing.T) {
	needsServer(t)

	client := aether.NewClient(baseURL)
	actorID := uniqueID("inttest")

	_, err := client.RegisterActor(actorID, "worker", nil, nil)
	if err != nil {
		t.Fatalf("register actor failed: %v", err)
	}

	err = client.Heartbeat(actorID)
	if err != nil {
		t.Fatalf("heartbeat failed: %v", err)
	}
}

func TestIntegrationSetGetState(t *testing.T) {
	needsServer(t)

	client := aether.NewClient(baseURL)
	actorID := uniqueID("inttest")

	client.RegisterActor(actorID, "worker", nil, nil)

	entry, err := client.SetState(actorID, "counter", 42, nil)
	if err != nil {
		t.Fatalf("set state failed: %v", err)
	}
	if entry.Version < 1 {
		t.Errorf("expected version >= 1, got %d", entry.Version)
	}

	got, err := client.GetState(actorID, "counter")
	if err != nil {
		t.Fatalf("get state failed: %v", err)
	}
	if got.Value != float64(42) {
		t.Errorf("expected value 42, got %v", got.Value)
	}
}

func TestIntegrationGetMissingState(t *testing.T) {
	needsServer(t)

	client := aether.NewClient(baseURL)
	actorID := uniqueID("inttest")

	client.RegisterActor(actorID, "worker", nil, nil)

	got, err := client.GetState(actorID, "nonexistent")
	if err != nil {
		t.Fatalf("get missing state should not error: %v", err)
	}
	if got != nil {
		t.Error("expected nil for missing state key")
	}
}

func TestIntegrationDeleteState(t *testing.T) {
	needsServer(t)

	client := aether.NewClient(baseURL)
	actorID := uniqueID("inttest")

	client.RegisterActor(actorID, "worker", nil, nil)
	client.SetState(actorID, "temp", "data", nil)

	err := client.DeleteState(actorID, "temp")
	if err != nil {
		t.Fatalf("delete state failed: %v", err)
	}
}

func TestIntegrationGetAllState(t *testing.T) {
	needsServer(t)

	client := aether.NewClient(baseURL)
	actorID := uniqueID("inttest")

	client.RegisterActor(actorID, "worker", nil, nil)
	client.SetState(actorID, "a", 1, nil)
	client.SetState(actorID, "b", 2, nil)

	resp, err := client.GetAllState(actorID)
	if err != nil {
		t.Fatalf("get all state failed: %v", err)
	}
	if len(resp.State) < 2 {
		t.Errorf("expected >= 2 state keys, got %d", len(resp.State))
	}
}

func TestIntegrationStateVersionIncrements(t *testing.T) {
	needsServer(t)

	client := aether.NewClient(baseURL)
	actorID := uniqueID("inttest")

	client.RegisterActor(actorID, "worker", nil, nil)

	e1, _ := client.SetState(actorID, "counter", 1, nil)
	e2, _ := client.SetState(actorID, "counter", 2, nil)

	if e2.Version <= e1.Version {
		t.Errorf("expected version to increment: %d -> %d", e1.Version, e2.Version)
	}
}

func TestIntegrationClusterInfo(t *testing.T) {
	needsServer(t)

	httpClient := &http.Client{Timeout: 5 * time.Second}
	resp, err := httpClient.Get(baseURL + "/cluster/info")
	if err != nil {
		t.Skipf("Cluster endpoint not available: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode == 404 {
		t.Skip("Cluster endpoints not available on this server")
	}
	if resp.StatusCode != 200 {
		t.Fatalf("expected 200, got %d", resp.StatusCode)
	}

	var data map[string]any
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		t.Fatalf("failed to decode cluster info: %v", err)
	}

	if _, ok := data["node_id"]; !ok {
		if _, ok := data["cluster_enabled"]; !ok {
			if _, ok := data["status"]; !ok {
				t.Error("expected node_id, cluster_enabled, or status in cluster info")
			}
		}
	}
}

func TestIntegrationClusterNodes(t *testing.T) {
	needsServer(t)

	httpClient := &http.Client{Timeout: 5 * time.Second}
	resp, err := httpClient.Get(baseURL + "/cluster/nodes")
	if err != nil {
		t.Skipf("Cluster endpoint not available: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode == 404 {
		t.Skip("Cluster endpoints not available on this server")
	}
	if resp.StatusCode != 200 {
		t.Fatalf("expected 200, got %d", resp.StatusCode)
	}
}
